Add Physics action for linear FV fluid energy equation

GiudGiud opened the pull request:
Contributes:
- a physics for linearFV + fluid energy
- add support for DumpObjectsProblem + multi-system OR linear system(s) and complex executioners (for example SIMPLE performs additional items that require the problem to be well setup, which the DOP is not)
this last item will be really nice for developing further physics with complex executioners / with future executors
I dont think it will be enough for debugging THM physics yet

GiudGiud committed:
Add a base class for fluid heat transfer physics
refs #29175

GiudGiud committed:
Add a linear FV physics class for energy transfer
refs #29175

GiudGiud committed:
Add testing for the linear energy transfer physics
refs #29175

GiudGiud committed:
Add support for creating outlet BCs for fluid energy physics

GiudGiud committed:
Add documentation for new physics

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2575747/), step Docs: sync website on 02ea865 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/29177/site)

This comment will be updated on new commits.

GiudGiud committed:
Add multi-system and linear system support to DumpObjectProblem

GiudGiud committed:
Output objects in a different manner: use a task at the end of the setup
phase rather than relying on the problem solve calls

GiudGiud committed:
Add boolean for two term BCs for momentum outflow bcs

GiudGiud committed:
Handle specific heat better in kernel and physics

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2577508/) on e552ba7 : invalidated by @loganharbour

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2577555/), step Generate coverage on e552ba7 wanted to post the following:


## Framework coverage

No coverage report(s) were found for the base commit 98dcedab1e.
[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29177/coverage/framework)


## Modules coverage

No coverage report(s) were found for the base commit 98dcedab1e.

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/29177/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented:
Testing was green pre-rebase.
@freiler want to take a look?

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2582624/), step Verify coverage on 0c5907e wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `electromagnetics` coverage rate (required: 94.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 85.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 85.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 81.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_transfer` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 85.0%)
- Failed to generate `misc` coverage rate (required: 32.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `peridynamics` coverage rate (required: 77.0%)
- Failed to generate `phase_field` coverage rate (required: 85.0%)
- Failed to generate `porous_flow` coverage rate (required: 95.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `rdg` coverage rate (required: 63.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `scalar_transport` coverage rate (required: 81.0%)
- Failed to generate `solid_mechanics` coverage rate (required: 84.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 88.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

lindsayad commented on framework/include/systems/DumpObjectsLinearSystem.h:
```diff
@@ -0,0 +1,37 @@
+#pragma once
+
+#include "LinearSystem.h"
+#include "libmesh/nonlinear_solver.h"
```
```suggestion
```

lindsayad commented on framework/src/actions/DumpObjectsAction.C:
```diff
@@ -0,0 +1,30 @@
+void
+DumpObjectsAction::act()
+{
+  if (!dynamic_cast<DumpObjectsProblem *>(_problem.get()))
+    return;
+  dynamic_cast<DumpObjectsProblem *>(_problem.get())->printObjects();
```
```suggestion
  if (auto * const dop = dynamic_cast<DumpObjectsProblem *>(_problem.get()); dop)
    dop->printObjects();
```

lindsayad commented on framework/src/problems/DumpObjectsProblem.C:
```diff
@@ -30,24 +33,57 @@ DumpObjectsProblem::validParams()
   // Create extra solution vectors if any
   createTagSolutions();
+
+  // Add an action to call printObjects at the end of the action/tasks phase
+  // NOTE: We previously relied on problem.solve() but some executioners (SIMPLE in NavierStokes) do
+  // not support this
+  auto action_params = _app.getActionFactory().getValidParams("DumpObjectsAction");
+  action_params.applyParameters(parameters);
+  // action_params.finalize("");
+  auto dump_objects_action =
+      _app.getActionFactory().create("DumpObjectsAction", "dump_objects", action_params);
+  _app.actionWarehouse().addActionBlock(dump_objects_action);
```
It seems a little weird to have the action added by the `MooseObject`. You already have the `dynamic_cast` check in the action. Why not just always add the action then and don't worry about having this code?

At a minimum if this problem adds the action, we should have no need for the `dynamic_cast` check in the action (could assert instead)

GiudGiud replied:
works for me

GiudGiud replied:
I like creating the action here so I ll remove the cast. 
I think the "Problem" pertains to the setup as well and adding an action is a setup action so it should be fine

GiudGiud replied:
ugh now I realize I redid all this work when working on the tutorial PR because I needed the DumpObjectsProblem to work there too. We can't wait 6 weeks to merge my PR, I have a two week memory

lindsayad commented on framework/src/problems/DumpObjectsProblem.C:
```diff
@@ -143,6 +179,18 @@ DumpObjectsProblem::dumpVariableHelper(const std::string & system,
 void
 DumpObjectsProblem::solve(unsigned int)
+{
+  printObjects();
+}
+
+void
+DumpObjectsProblem::solveLinearSystem(unsigned int, const Moose::PetscSupport::PetscOptions *)
+{
+  printObjects();
```
Do we need these calls anymore if the action is doing the `printObjects` call?

GiudGiud replied:
A derived class (not sure anyone had one) could be using the old solve to print

lindsayad replied:
Yea I'm confident we don't have anyone deriving from `DumpObjectsProblem`

GiudGiud replied:
we still need to override solve imo to avoid trying to solve without initialSetup called. but should be fine now

GiudGiud committed:
Address Alex-s review

- more compact syntax
- remove extra header
- remove any form of solve from DumpObjectsProblem but dont both printing again

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

lindsayad commented on framework/src/actions/DumpObjectsAction.C:
```diff
@@ -0,0 +1,31 @@
+void
+DumpObjectsAction::act()
+{
+  mooseAssert(dynamic_cast<DumpObjectsProblem *>(_problem.get()),
+              "Problem should be DumpObjectProblem");
+  if (auto * const dop = dynamic_cast<DumpObjectsProblem *>(_problem.get()); dop)
+    dop->printObjects();
```
```suggestion
  auto * const dop = dynamic_cast<DumpObjectsProblem *>(_problem.get());
  mooseAssert(dop, "Problem should be DumpObjectProblem");
  dop->printObjects();
```

lindsayad replied:
sorry, I think it can be a little simpler now

GiudGiud replied:
I ll merge it in the NS review

lindsayad reviewed (commented):
I think the framework part looks good to me. I can be the NS reviewer as well unless @grmnptr wants to take it

GiudGiud commented:
@freiler want to do the first pass? We need more NS people familiar with our de factor action syntax imo

GiudGiud commented on modules/navier_stokes/src/physics/WCNSFVFluidHeatTransferPhysicsBase.C:
```diff
@@ -0,0 +1,254 @@
+  params.set<MooseFunctorName>("temperature") = _fluid_temperature_name;
+
+  getProblem().addMaterial(
+      "INSFVEnthalpyFunctorMaterial", prefix() + "ins_enthalpy_material", params);
```
I think we can put in Ramiro's functor material here once it's merged. We can just add a boolean "solve_for_enthalpy" to make the change of variable too. 
https://github.com/idaholab/moose/pull/29532/files#diff-230652c10695051399dccb2366c0e4d1b6ba10acd6322d9c3e06b733f5608edf

@grmnptr can you please push this PR forward?

grmnptr replied:
Yeah I reviewed it yesterday. I'll ask if they can address the comments or if I should do it.

GiudGiud replied:
I meant this PR as in this one. I think the follow up post Ramiro's can be done in a different PR

grmnptr reviewed (changes requested):
Very nice!

grmnptr commented on modules/navier_stokes/src/physics/WCNSFVFluidHeatTransferPhysicsBase.C:
```diff
@@ -0,0 +1,254 @@
+}
+
+unsigned short
+WCNSFVFluidHeatTransferPhysicsBase::getNumberAlgebraicGhostingLayersNeeded() const
```
This is 1 for the linear version.

GiudGiud replied:
yup it's already like this
```
unsigned short
WCNSLinearFVFlowPhysics::getNumberAlgebraicGhostingLayersNeeded() const
{
  return 1;
}
```

grmnptr replied:
Nice! Sorry, I missed this.

grmnptr reviewed (commented):
Oh I forgot this one.

GiudGiud committed:
Address Peter's review

- improve parameter description
- format/bracket fix
- improve doco heat->advection
- list functors being create in doco

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>
Co-authored-by: Peter German <31662443+grmnptr@users.noreply.github.com>

GiudGiud commented on modules/navier_stokes/doc/content/source/physics/WCNSLinearFVFluidHeatTransferPhysics.md:
```diff
@@ -0,0 +1,59 @@
+- [LinearFVVolumetricHeatTransfer.md] for the volumetric ambient convection term, if present
+
+
+## Automatically defined variables
```
done

GiudGiud commented on modules/navier_stokes/include/physics/WCNSFVFluidHeatTransferPhysicsBase.h:
```diff
@@ -0,0 +1,117 @@
+
+  /// Process thermal conductivity (multiple functor input options are available).
+  /// Return true if we have vector thermal conductivity and false if scalar
+  bool processThermalConductivity();
```
i like process because it's what it actually does. The bool is just a conseuqnece of the processing. I think you might have named this process back in the Action times

grmnptr replied:
Sounds good!

GiudGiud commented on modules/navier_stokes/include/physics/WCNSFVFluidHeatTransferPhysicsBase.h:
```diff
@@ -0,0 +1,117 @@
+   * If the material properties are not constant, some of these can be used for
+   * weakly-compressible simulations as well.
+   */
+  virtual void addINSEnergyTimeKernels() = 0;
+  virtual void addWCNSEnergyTimeKernels() = 0;
+  virtual void addINSEnergyHeatConductionKernels() = 0;
+  virtual void addINSEnergyAdvectionKernels() = 0;
+  virtual void addINSEnergyAmbientConvection() = 0;
+  virtual void addINSEnergyExternalHeatSource() = 0;
```
as in get rid of INS?

GiudGiud replied:
I could do this post your PR when adding the transient support to the Physics if you agree. I ll need to add the energy kernel routine for linear FV then

grmnptr replied:
Sounds good!

GiudGiud commented on modules/navier_stokes/include/physics/WCNSFVFluidHeatTransferPhysicsBase.h:
```diff
@@ -0,0 +1,117 @@
+  virtual void addINSEnergyAmbientConvection() = 0;
+  virtual void addINSEnergyExternalHeatSource() = 0;
+
+  /// Functions adding boundary conditions for the incompressible simulation.
```
some of the kernels are different so that's why.
There is a compressibilty parameter to the NSFVPhysics you set

grmnptr replied:
Yeah I know, but in theory the two formulations are the same from the linear fv side at least. If you use Newton's method you have slight differences for sure. Just brainstorming here how to make life simpler :D.

GiudGiud commented on modules/navier_stokes/src/physics/WCNSLinearFVFluidHeatTransferPhysics.C:
```diff
@@ -0,0 +1,279 @@
+    const auto var_type = "MooseLinearVariableFVReal";
+    auto params = getFactory().getValidParams(var_type);
+    assignBlocks(params, _blocks);
+    params.set<std::vector<Real>>("scaling") = {getParam<Real>("energy_scaling")};
```
but ignoring it would be assuming the design of the executioner. A future executioner might do something else

grmnptr replied:
Yeah that is true.

GiudGiud commented:
Rebased on next to confirm no issue with the new INSFVEnthalpyFunctorMaterial setup. 
Still have not made using 'h' the default

grmnptr reviewed (approved):
Looks good to me! Failures are unrelated.
